### DISCUSSION TITLE
Best way to override Controllers and actions from /vendor/statamic/cms (routes/cp.php) ?

### DISCUSSION THREAD

**Discussion by sarahessex** [statamic/cms#5411] [Q&A] [answered]
Today is my first day using Statamic and I am trying to override the `validateUniqueUri` function in `Statamic\Http\Controllers\CP\Collections\EntriesController` (the function is called within `EntriesController@create` and `EntriesController@update`) for reasons specific to my project. Haven't been able to find much documentation on how to do this so not sure of the best approach.

To attempt to do this, I have created a private addon with a Controller class extending `Statamic\Http\Controllers\CP\Collections\EntriesController` and in the addon ServiceProvider I have successfully created a _new_ route using `Statamic::pushCpRoutes`, but I don't seem to be able to override existing routes and their actions.

My current code...

**addons/addonvendor/addonid/src/Http/Controllers/EntriesController.php**

```
<?php

namespace AddonVendor\AddonId\Http\Controllers;

use Illuminate\Http\Request;
use Illuminate\Validation\ValidationException;
use Statamic\Contracts\Entries\Entry as EntryContract;
use Statamic\CP\Breadcrumbs;
use Statamic\Exceptions\BlueprintNotFoundException;
use Statamic\Facades\Asset;
use Statamic\Facades\Entry;
use Statamic\Facades\Site;
use Statamic\Facades\User;
use Statamic\Http\Controllers\CP\Collections\EntriesController as StatamicEntriesController;
use Statamic\Http\Requests\FilteredRequest;
use Statamic\Http\Resources\CP\Entries\Entries;
use Statamic\Http\Resources\CP\Entries\Entry as EntryResource;
use Statamic\Query\Scopes\Filters\Concerns\QueriesFilters;
use Statamic\Support\Arr;
use Statamic\Support\Str;

class EntriesController extends StatamicEntriesController
{
    public function foo(Request $request)
    {
        return view('foo'); // this is working
    }

    public function update(Request $request, $collection, $entry)
    {
        // this function is not running, default Statamic\Http\Controllers\CP\Collections\EntriesController@update is still being used
        // ...
        $this->validateUniqueUri($entry, $tree ?? null, $parent ?? null);
        // ...
    }

    public function create(Request $request, $collection, $site)
    {
        // ... this function is not running, default Statamic\Http\Controllers\CP\Collections\EntriesController@create is still being used
        // ...
        $this->validateUniqueUri($entry, $tree ?? null, $parent ?? null);
        // ...
    }

    private function validateUniqueUri($entry, $tree, $parent)
    {
        // this is the function I want to override
    }
}

```

**addons/addonvendor/addonid/src/ServiceProvider.php**
```
<?php

namespace AddonVendor\AddonId;

use Statamic\Providers\AddonServiceProvider;
use Illuminate\Support\Facades\Route;
use Statamic\Statamic;

class ServiceProvider extends AddonServiceProvider
{

    public function bootAddon()
    {
        Statamic::pushCpRoutes(function () {

            // this new route is working
            Route::namespace('\AddonVendor\AddonId\Http\Controllers')->group(function () {
                Route::get('foobar', 'EntriesController@foo');
            });

            // none of these attempts to override the existing routes are working
            Route::namespace('\AddonVendor\AddonId\Http\Controllers')->group(function () {
                Route::group(['prefix' => 'collections/{collection}/entries'], function () {
                    Route::get('create/{site}', 'EntriesController@create')->name('collections.entries.create');
                    Route::group(['prefix' => '{entry}'], function () {
                        Route::patch('/', 'EntriesController@update')->name('collections.entries.update');
                    });
                });
            });
        });

    }
}
```

The `cp/foobar` route I added in the above ServiceProvider is working, but the routes `collections.entries.create` and `collections.entries.update` are not being overridden and still using the original Controller from `/vendor/statamic/` that I am trying to override. I'm not getting any errors, it just seems to be that what I am trying to do isn't registering at all. What is wrong with my code?

Alternatively, if anyone has any suggestions for how I can override the `validateUniqueUri` function a different way I am all ears! Thanks.

**Comment by jasonvarga** [comment, marked as answer]
Welcome to Statamic!

Maybe it's easier to know why you're wanting to override the validation in the first place?

**Reply by sarahessex** [reply]
Hi! Thank you. It's a bit convoluted but the reason is... the client has different 'sectors' to their website, there might be multiple entries that use the same URI (slug) as they will live under different routes. The URL structure we want is `website.com/{sector}/products/{product-uri}` eg. `website.com/sector-one/products/product-one` and `website.com/sector-two/products/product-one`.

They want the **content** of these product pages with the same slug to be different (as they have different content speaking to different sectors), but they want the products to lie under the same product identifiers/slug. So we've set up multiple product entries with the intention of them having the same slug, and the client will tick which Sector (which I have set up as a Taxonomy) that product applies to and assign whatever content they want to each product.

The frontend  routing is working ok to find the correct sector and product for the given URL, after we've added some custom routing to routes/web.php, but I'm getting the error "This URI has already been taken." when saving the entries that apply to different sectors (because they do, indeed, have the same URI!). So I am trying to override the `validateUniqueUri` function to check the taxonomy and only throw the duplicate URI error if it's under the same sector/taxonomy, but not for a product under a different sector/taxonomy.

There are other collections which will apply to the sectors too (not just products), such as blog articles, FAQs, etc. The idea being that the client can assign the entries to one or multiple sectors.

Hope that makes some sense?!

**Reply by jasonvarga** [reply]
Sounds like the URIs should still be unique, no? The `{sector}` would be part of the URI.

How are you defining a sector? Is that a taxonomy?

How about you just don't give the collection a `route`. You can still give the entry `slug`s - they just wont have URLs. Since you're rolling your own routing, that should work. If the Collection doesn't have a route, Statamic won't try to validate the URI.

**Reply by sarahessex** [reply]
Ok, that works. Thank you. In the end, I've set the collection route to be `/{{if segments}}{{segments:0}}/{{/if}}products/{{slug}}` and it seems to give me the desired result. I was thinking URI and slug were one and the same, didn't realise URI validated the entire route.

Regardless of that solution for my use case though, it would still be good to know or get a pointer on how to override the controller for default Statamic routes, if possible? What was I missing?

**Reply by jasonvarga** [reply]
Overriding one of our controllers isn't really something I would ever suggest. But if you really have to, you could probably re-bind it in the container. `$this->app->bind(EntriesController:class, YourCustomController::class);` But please don't.

**Reply by sarahessex** [reply]
Ok, understood! Thanks
